### ISSUE TITLE
multivariate_normal.log_prob is slow

### ISSUE THREAD

**Issue by t-vi** [pytorch/pytorch#17206]
## 🐛 Bug

On the discussion forums, user Al3x reported a severe performance regression - from subsecond to almost 9 minutes:
https://discuss.pytorch.org/t/going-from-0-4-0-to-1-0-0-changes-code-runtime-from-000-35-to-045-63/37200/15
The problematic bit is MultivariateNormal. I tracked this down to `_batch_mahalanobis` and it's use  of `_batch_trtrs_lower`. In 0.4 it used (a list comprehension over non-batched, because we didn't have it) torch.inverse instead of `trtrs`.

## To Reproduce

```
mu = torch.randn(8)
sigma = torch.eye(8).expand(6, 1, 8, 8).contiguous().requires_grad_()
x_repeat = torch.randn(6,8000,8)
pdf = MultivariateNormal(mu, sigma).log_prob(x_repeat).t()
pdf.sum().backward()
```

More specific code is also in the forum thread.

## Expected behavior

The old speed!

## Environment

Masterish PyTorch and comparison to 0.4.1.

## Additional context

I'm undecided whether it's `_batch_mahalanobis`' or `trtrs`'s fault. Probably we should have a triangular solver primitive without the funny extra copy return and a batch version, too.

CC some distribution gurus @fehiepsi and @fritzo and batch-linear-algebra hero @vishwakftw

**Comment by fehiepsi** [comment]
@t-vi Thanks for notifying me! I think that it is my fault when trying to use `trtrs` instead of `inverse`. The slowdown seems lie at [this line](https://github.com/pytorch/pytorch/blob/master/torch/distributions/multivariate_normal.py#L43) where we expand `bL` batch shape to match `bx` batch shape, then do triangle solve with "expanded" bL. If we use `inverse`, then after taking the inverse of `bL`, we just simply do `matmul` with `bx` to get the result. So it will be much faster in the above example.

I'll sketch out a solution which I have in mind for further discussion.

**Comment by fehiepsi** [comment]
I make a [gist](https://gist.github.com/fehiepsi/da4f45ad00d450ee74ffe6e7809c368a) to show my proposed solution. We can see a huge speed up with @t-vi 's example.

Basically, assume `bL.shape = (i, j, n, n)` and `bx.shape = (K, I, J, n)`, we first reshape `bx` such that
`bx.shape = (K, I / i, i, J / j, j, n)`. Then we move "to-be-expaned" dims to the left by permuting such that
`bx.shape = (K, I / i, J / j, i, j, n)`. Now, we don't need to expand `bL` because the batch_shape (i, j) is matched. After using triangle_solve and getting a result with shape `(K, I / i, J / j, i, j)`, we permute the result to make it have shape `(K, I / i, i, J / j, j)`. A final reshape operator will give shape `(K, I, J)` for the result.

Though this solves the topic's issue, there might be a better solution without having to manipulate reshape+permute operators like this. I'm looking forward to see others' opinions.

**Comment by t-vi** [comment]
Thank you. Your analysis is spot on. That, on the other hand, means that it's not using that using `trtrs` itself is a problem.

In terms of solution, I'd think that getting a good, batch-capable triangular_solve in ATen would be ideal.
For the proper broadcasting:
- trtrs takes c right hand sides in `(n,c)` shape (ugh).
- Prepend singleton dimensions to `bx` until `bx.dim()+1 >= bL.dim()`.
- Move the first `bx.dim() + 1 - bL.dim() ` dimensions of `bx` to the end (after `n`).
- Eliminate any singleton/stride 0 dimensions from `bL` while moving the corresponding dims in`bx` to the right.
- Then each `(n,n)` block in `bL` would correspond to `(n, ...)` bx and one could apply `trtrs` to solve that.
- Finally we need to clean up all the permutation mess. :)

**Comment by fehiepsi** [comment]
@t-vi Thanks! I forgot that PyTorch only supports broadcasting for `j=1` or `j=J`, so we don't have to `reshape`. :)

**Comment by soumith** [comment]
@fehiepsi go for it, it looks like no one else is jumping in with comments.

**Comment by t-vi** [comment]
This is fixed by #17294. Thank you, @fehiepsi !
